feat/back/api-endpoints

yandrinsky opened the pull request:
- create task
 - start solution
 - get current solution
 - commit

yandrinsky committed:
add the following endpoints:
 - create task
 - start solution
 - get current solution
 - commit

yandrinsky committed:
add the following endpoints:
 - new try
 - switch try
 - switch head frame

yandrinsky committed:
hotfix:
 - forbade same name tasks
 - allow creation of a task only for admins and creators
 - in start solution endpoint: check if task doesn't exist
 - optimisation with Promise.all

kyzinatra commented on packages/backend/api/controllers/api/commit/commit.controller.ts:
```diff
@@ -0,0 +1,56 @@
+    try {
+        let res = await Promise.all([
+            Solution.findOne({ ownerId: req.user?._id, taskId }),
+            Try.findOne({ _id: tryId }),
```
Ты же говорил, что тебя бесит слово Solution хахаха

kyzinatra commented on packages/backend/api/controllers/api/commit/commit.controller.ts:
```diff
@@ -0,0 +1,56 @@
+
+        solution = res[0];
+        currentTry = res[1];
+        parentFrame = res[2];
```
Почему не декструкторизацией

yandrinsky replied:
Поправлю

kyzinatra commented on packages/backend/api/controllers/api/get-current-solution/get-current-solution.controller.ts:
```diff
@@ -0,0 +1,50 @@
+    try {
+        solution = await Solution.findOne({ ownerId: req.user?._id, taskId });
+    } catch (e) {
+        return resp.status(SERVER_ERRORS.BD_ERROR.code).json(SERVER_ERRORS.BD_ERROR);
```
че на кринж. Используй .catch(). Зачем разводить этот try catch огромный

yandrinsky replied:
Я через catch return не сделаю, чтобы дальше код не исполнялся

kyzinatra commented on packages/backend/api/controllers/api/start-solution/start-solution.controller.ts:
```diff
@@ -0,0 +1,42 @@
+
+    try {
+        solution = await Solution.findOne({ ownerId: req.user?._id, taskId });
+    } catch (e) {
```
кринж

yandrinsky replied:
Чел знакомится с конструкциями языка

kyzinatra commented on packages/backend/bd/schemas/frame.schema.ts:
```diff
@@ -0,0 +1,16 @@
+        comment: { type: String, default: null }
+    },
+    { _id: false }
+);
```
А ты индекс для _id поставил?

yandrinsky replied:
Я поставил свой _id, чтобы он был строкой. У них всратый ObjectId, с ним неудобно работать

kyzinatra replied:
Ну окей, индекс для него поставь

kyzinatra commented on packages/backend/bd/schemas/frame.schema.ts:
```diff
@@ -0,0 +1,16 @@
+    { _id: false }
+);
+
+export const Frame = model('Frame', frameSchema);
```
С маленькой буквы model('frame')

yandrinsky replied:
В доке с большой

kyzinatra replied:
Ладно принимается. Просто все равно с маленькой будет

kyzinatra commented on packages/backend/bd/schemas/solution.schema.ts:
```diff
@@ -0,0 +1,19 @@
+        _id: { type: String, default: randomUUID }
+    },
+    { _id: false }
+);
```
А индекс где

yandrinsky replied:
_id: false означает, что _id не будет автоматическим, а будет по моим правилам. Он на 2 строчки выше

kyzinatra replied:
Индекс где?

kyzinatra commented on packages/backend/bd/schemas/solution.schema.ts:
```diff
@@ -0,0 +1,19 @@
+    { _id: false }
+);
+
+export const Solution = model('Solution', solutionSchema);
```
с маленько буквы

yandrinsky replied:
В доке с большой

kyzinatra commented on packages/backend/bd/schemas/try.schema.ts:
```diff
@@ -0,0 +1,19 @@
+    },
+    { _id: false }
+);
+
```
По аналогии

yandrinsky replied:
По аналогии

kyzinatra replied:
По аналогии добавь индекс

kyzinatra commented on packages/backend/domain/middleware/error-bounding-middleware.ts:
```diff
@@ -0,0 +1,15 @@
+    next: NextFunction
+): Promise<void> => {
+    console.error('errorBoundingMiddleware', err.stack);
+    resp.status(SERVER_ERRORS.SOMETHING_WENT_WRONG.code).json(SERVER_ERRORS.SOMETHING_WENT_WRONG);
```
че это. Если хочешь логов скачай библу для создания и сохранения логов. А из консоли убери. Они useless

kyzinatra commented on packages/backend/domain/utils/get-path-from-tree-from-end.ts:
```diff
@@ -0,0 +1,39 @@
+
+    const step = (
+        node: IDoublyConnectedTree<T['data'] & { _id: string }>,
+        id: string
```
Тип лучше вынести отдельно. Очень длинный и используется 3 раза подряд

kyzinatra commented on packages/backend/domain/utils/get-path-from-tree-from-end.ts:
```diff
@@ -0,0 +1,39 @@
+
+        if (node.parent) {
+            //@ts-ignore
+            step(node.parent, node.parent.data._id);
```
Чееел. Че это за кринжатина. Быстро убирай пока никто не увидел

kyzinatra commented on packages/backend/domain/utils/get-path-from-tree-from-end.ts:
```diff
@@ -0,0 +1,39 @@
+    step(
+        treeForwardTraversal(tree, ({ _id }) => _id === startId) as IDoublyConnectedTree<
+            T['data'] & { _id: string }
+        >,
```
Да, вынеси этот тип. Он очень большой. А лучше, чтобы вообще не надо было постоянно писать as.

kyzinatra commented on packages/backend/domain/utils/make-tree-doubly-connected.ts:
```diff
@@ -0,0 +1,21 @@
+    initialTree: T
+): IDoublyConnectedTree<T> => {
+    const tree = JSON.parse(JSON.stringify(initialTree));
+
```
Быстро убирай этот кринж. Для копирования объекта используй `structuredClone()`

yandrinsky replied:
Кайф, не знал о нем

kyzinatra commented on packages/backend/domain/utils/make-tree-doubly-connected.ts:
```diff
@@ -0,0 +1,21 @@
+
+    const step = (node: ITree<T>, parentNode: ITree<T> | null) => {
+        //@ts-ignore
+        node.parent = parentNode;
```
Кринж

kyzinatra commented on packages/backend/domain/utils/tree-forward-traversal.ts:
```diff
@@ -0,0 +1,33 @@
+export function treeForwardTraversal<T extends IDoublyConnectedTree<T['data']>>(
+    tree: T,
+    checkCallback: (tree: T['data']) => boolean
+): IDoublyConnectedTree<T['data']> | void;
```
Так перегрузки не работают. Зачем. Почему не сделать объединение`extends ` типов и использовать их. В чем смысл того, что ты 3 раза повторил одно и тоже

kyzinatra reviewed (changes requested):
Кайфово. Переделывай :)

yandrinsky committed:
feat: add new end-points and some refactor
 - get-task
 - get-tasks-list
 - get-not-approved-tasks-list
 - approve-task
 - update-task

 add 2 utils functions to manage files (saveFile, removeFile)

 Now we can send img preview while creating a new task

yandrinsky committed:
feat: add new end-points and some refactor
 - update task source not finished

yandrinsky committed:
feat
